zcash_client_sqlite/src/wallet.rs

```diff
@@ -519,28 +521,264 @@ pub(crate) fn get_balance(
+
+                let incomplete_tree_start = conn
+                    .query_row(
+                        "SELECT sapling_commitment_tree_size
+                         FROM blocks
+                         WHERE height = :fully_scanned_height",
+                        named_params![":fully_scanned_height": u32::from(fully_scanned_height)],
+                        |row| row.get::<_, u32>(0),
+                    )
+                    .optional()?;
+
+                let incomplete_start: u64 = incomplete_tree_start.map_or_else(
+                    || conn.query_row(
+                        "SELECT start_position
+                         FROM v_sapling_shard_scan_state
+                         WHERE subtree_start_height <= :fully_scanned_height
+                         AND (subtree_end_height > :fully_scanned_height OR subtree_end_height IS NULL)", 
+                         named_params![":fully_scanned_height": u32::from(fully_scanned_height)], 
+                         |row| row.get::<_, u32>(0)
+                     ).optional()
+                    .map(|opt| opt.unwrap_or(0)),
+                    Ok
+                )?.into();
+
+                Ok(Some(Ratio::new(
+                    scanned,
+                    (tip.start_position - incomplete_start) + tip_notes
+                )))
+            } else {
+                // the chain tip shard is the only one
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// Returns the spendable balance for the account at the specified height.
+///
+/// This may be used to obtain a balance that ignores notes that have been detected so recently
+/// that they are not yet spendable, or for which it is not yet possible to construct witnesses.
+pub(crate) fn get_wallet_summary(
+    conn: &rusqlite::Connection,
+    min_confirmations: u32,
+) -> Result<Option<WalletSummary>, SqliteClientError> {
+    let chain_tip_height = match scan_queue_extrema(conn)? {
+        Some((_, max)) => max,
+        None => {
+            return Ok(None);
+        }
+    };
+
+    let birthday_height =
+        wallet_birthday(conn)?.expect("If a scan range exists, we know the wallet birthday.");
+
+    let fully_scanned_height =
+        block_fully_scanned(conn)?.map_or(birthday_height - 1, |m| m.block_height());
+
+    let sapling_scan_progress =
+        sapling_scan_progress(conn, fully_scanned_height, chain_tip_height)?;
+
+    let summary_height = chain_tip_height + 1 - min_confirmations;
+
+    // If any of the shard below the anchor (or chain tip if no anchor is specified) is unscanned
+    // above the wallet birthday height, none of our balance is currently spendable.
+    let any_spendable = conn
+        .query_row(
+            "SELECT 1 FROM v_sapling_shard_unscanned_ranges
+             WHERE :summary_height
+                BETWEEN subtree_start_height
+                AND IFNULL(subtree_end_height, :summary_height)
```

Comment:
If `min_confirmations` is zero, then the chain could be fully scanned up to and including a complete shard at `chain_tip_height`, and this query would return no results because `summary_height` would be one greater than `subtree_end_height`. This seems not to be the intent.